# Genomic Data - HumanEnhancersEnsembl

This is one of the genomic datasets taken from [here](https://github.com/ML-Bioinfo-CEITEC/genomic_benchmarks).
The classification task is evaluated using the _SeqRep_ package.

You can [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MIR-MU/seqrep/blob/main/examples/genomic_data/HumanEnhancersEnsembl.ipynb)
or
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/MIR-MU/seqrep/main?labpath=examples%2Fgenomic_data%2FHumanEnhancersEnsembl.ipynb).

## Install _SeqRep_ Package

In [1]:
!pip install seqrep

     |████████████████████████████████| 115 kB 7.3 MB/s 
     |████████████████████████████████| 303 kB 14.9 MB/s 
     |████████████████████████████████| 15.3 MB 442 kB/s 
  Created wheel for pandas-ta: filename=pandas_ta-0.3.14b0-py3-none-any.whl size=218923 sha256=24c47b486a44e3c4a3c1b6330dfcfcb000e9c47398544c88ae66d2944aca9631
  Stored in directory: /root/.cache/pip/wheels/0b/81/f0/cca85757840e4616a2c6b9fe12569d97d324c27cac60724c58
  Created wheel for ta: filename=ta-0.9.0-py3-none-any.whl size=28908 sha256=87415e4559c04d406783c5bd56b8377e348dd1fcf14f3b3a4370cecda6fb7f69
  Stored in directory: /root/.cache/pip/wheels/72/78/64/cc1c01506a1010a9845e9bd7c69333730f7174661228ea4f98
Successfully built pandas-ta ta
  Attempting uninstall: numpy
    Found existing installation: numpy 1.21.5
    Uninstalling numpy-1.21.5:
      Successfully uninstalled numpy-1.21.5
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully u

## Import Needed Packages

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.neural_network import MLPClassifier

!pip install icecream
from icecream import ic

from seqrep import *
from seqrep.feature_engineering import *
from seqrep.labeling import *
from seqrep.splitting import *
from seqrep.scaling import *
from seqrep.feature_reduction import *
from seqrep.evaluation import *
from seqrep.pipeline_evaluation import *

## Download Data

In [4]:
!pip install genomic-benchmarks

     |████████████████████████████████| 2.3 MB 10.4 MB/s 
     |████████████████████████████████| 596 kB 61.6 MB/s 
     |████████████████████████████████| 271 kB 59.5 MB/s 
     |████████████████████████████████| 94 kB 2.2 MB/s 
  Created wheel for genomic-benchmarks: filename=genomic_benchmarks-0.0.6-py3-none-any.whl size=17526 sha256=53565a8f0341c749751577e40c3d628a1ebae9d8b4e5f04c963f76f81a8d8f2d
  Stored in directory: /root/.cache/pip/wheels/a7/0f/e4/f723fa5070923dc495b63e02600faf64bf76dbdf422ebdeb66
Successfully built genomic-benchmarks
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [5]:
from genomic_benchmarks.data_check import list_datasets

list_datasets()

['human_nontata_promoters',
 'demo_human_or_worm',
 'human_enhancers_ensembl',
 'demo_coding_vs_intergenomic_seqs',
 'demo_mouse_enhancers',
 'human_enhancers_cohn']

In [6]:
from genomic_benchmarks.data_check import info

info("human_enhancers_ensembl", version=0)

Dataset `human_enhancers_ensembl` has 2 classes: negative, positive.

The lenght of genomic intervals ranges from 2 to 573, with average 268.8641324705183 and median 269.0.

Totally 154842 sequences have been found, 123872 for training and 30970 for testing.


train   test
negative  61936  15485
positive  61936  15485

In [20]:
for ds in list_datasets():
    info(ds, version=0)
    print("#######################")

Dataset `human_nontata_promoters` has 2 classes: negative, positive.

All lenghts of genomic intervals equals 251.

Totally 36131 sequences have been found, 27097 for training and 9034 for testing.
#######################
Dataset `demo_human_or_worm` has 2 classes: human, worm.

All lenghts of genomic intervals equals 200.

Totally 100000 sequences have been found, 75000 for training and 25000 for testing.
#######################
Dataset `human_enhancers_ensembl` has 2 classes: negative, positive.

The lenght of genomic intervals ranges from 2 to 573, with average 268.8641324705183 and median 269.0.

Totally 154842 sequences have been found, 123872 for training and 30970 for testing.
#######################
Dataset `demo_coding_vs_intergenomic_seqs` has 2 classes: coding_seqs, intergenomic_seqs.

All lenghts of genomic intervals equals 200.

Totally 100000 sequences have been found, 75000 for training and 25000 for testing.
#######################
Dataset `demo_mouse_enhancers` has 2 c

In [7]:
# !rm -rf ../root/.genomic_benchmarks/demo_mouse_enhancers

In [8]:
# from genomic_benchmarks.dataset_getters.pytorch_datasets import HumanEnhancersEnsembl

# X_train = HumanEnhancersEnsembl(split="train", version=0)
# X_test = HumanEnhancersEnsembl(split="test", version=0)

In [9]:
from genomic_benchmarks.loc2seq import download_dataset

download_dataset("human_enhancers_ensembl", version=0, use_cloud_cache=False)

/root/.genomic_benchmarks/fasta/Homo_sapiens.GRCh38.dna.toplevel.fa.gz: 0.00B [00:00, ?B/s]

  0%|          | 0/24 [00:00<?, ?it/s]

PosixPath('/root/.genomic_benchmarks/human_enhancers_ensembl')

In [10]:
!ls -al ../root/.genomic_benchmarks/human_enhancers_ensembl

total 16
drwxr-xr-x 4 root root 4096 Feb 24 09:17 .
drwxr-xr-x 4 root root 4096 Feb 24 09:17 ..
drwxr-xr-x 4 root root 4096 Feb 24 09:17 test
drwxr-xr-x 4 root root 4096 Feb 24 09:17 train


In [11]:
from genomic_benchmarks.dataset_getters.pytorch_datasets import HumanEnhancersEnsembl

X_train = HumanEnhancersEnsembl(split="train", version=0)
X_test = HumanEnhancersEnsembl(split="test", version=0)

y_train = pd.Series([y for _, y in X_train])
X_train = pd.DataFrame([x for x, _ in X_train], columns=["genom"])
y_test = pd.Series([y for _, y in X_test])
X_test = pd.DataFrame([x for x, _ in X_test], columns=["genom"])

In [ ]:
# from genomic_benchmarks.dataset_getters.pytorch_datasets import DemoMouseEnhancers

# X_train = DemoMouseEnhancers(split="train", version=0)
# X_test = DemoMouseEnhancers(split="test", version=0)

# y_train = pd.Series([y for _, y in X_train])
# X_train = pd.DataFrame([x for x, _ in X_train], columns=["genom"])
# y_test = pd.Series([y for _, y in X_test])
# X_test = pd.DataFrame([x for x, _ in X_test], columns=["genom"])

In [12]:
## Random shuffling
import numpy as np

idx = np.random.permutation(len(y_train))
X_train = X_train.iloc[idx, :].reset_index(drop=True)
n = len(y_train) // 2
X_train = X_train[:n]
y_train = y_train.iloc[idx].reset_index(drop=True)
y_train = y_train[:n]

idx = np.random.permutation(len(y_test))
X_test = X_test.iloc[idx, :].reset_index(drop=True)
y_test = y_test.iloc[idx].reset_index(drop=True)

X_train.join(pd.DataFrame(y_train, columns=["label"]))

genom  label
0      GTAGTTGGGAGTAGGTGCACACCACCATGCCCAGCGAATTTTTTGT...      1
1                                     CCAAAAAAAAAAAAAAAA      0
2      GGCTTTTCTCCAGAAGCAATAAAAACTGGAAATTAGCTCTGTCCCC...      0
3      CTTGATTGAATTCAGATTGTAAATTTTATGAGGGAGGGGTAAGACT...      1
4      CAGATAGAGAAGAATGGAATTACTGCTGCTGTATATATAACAGGGG...      0
...                                                  ...    ...
41285  GAAAAATATTGACATAGAACCTTATATTTGGAAAAGTATTTACATA...      1
41286  GATCTAAAGTTAAAAATAATACAAGTGGATTCAAATTCATGGCTAA...      1
41287  AGTTAAACTCTCTGATTCTCAGTTTTCTCACCAATAAAATTTGAGA...      1
41288  AGAGGGTAAGGGGGAAGGATGGCCTGGCATTAAGAGGGAGTTCCCT...      1
41289  TGCCTCCACTTCCACAGAAGACAGCCAGGCTGCGCGGCAGAGCCTG...      0

[41290 rows x 2 columns]

In [13]:
y_train.value_counts()

1    20730
0    20560
dtype: int64

## Run Pipeline Evaluation

In [14]:
# This DataFrame collects the results of various runs for comparison.

# Uncomment following line if you want to clear the DataFrame with the results.
# del results_for_comparison

try:
    results_for_comparison
except NameError:
    print("Create new empty DataFrame.")
    results_for_comparison = pd.DataFrame()
else:
    print("DataFrame already exist!")

Create new empty DataFrame.


In [15]:
class SubstringsExtractor(FeatureExtractor):
    def __init__(
        self,
        substrings: List,
        columns_to_apply: Union[str, List[str]] = None,
        return_original_columns: bool = False,
        normalize: bool = True,
        verbose: bool = True,
        inplace: bool = False,
    ):
        self.substrings = substrings
        self.columns_to_apply = columns_to_apply
        self.return_original_columns = return_original_columns
        self.normalize = normalize
        self.verbose = verbose
        self.inplace = inplace

    def fit(self, X, y=None):
        if self.verbose:
            print(f"\tNumber of substrings BEFORE fit: {len(self.substrings)}")
        if not self.columns_to_apply:
            self.columns_to_apply = X.columns
        if isinstance(self.columns_to_apply, str):
            columns_to_apply = [self.columns_to_apply]

        tmp = ""
        for c in self.columns_to_apply:
            tmp += "@".join(X[c])
        self.substrings = [
            s
            for s in tqdm(
                self.substrings, leave=False, desc="Fitting SubstringsExtractor"
            )
            if s in tmp
        ]
        del tmp

        if self.verbose:
            print(f"\tNumber of substrings AFTER fit:  {len(self.substrings)}")
        return self

    def transform(self, X):
        if not self.inplace:
            X = X.copy()
        for column in tqdm(
            self.columns_to_apply,
            leave=False,
            desc="Transforming SubstringsExtractor - columns",
        ):
            col_pref = column + "_" if len(self.columns_to_apply) > 1 else ""
            for substr in tqdm(
                self.substrings,
                leave=False,
                desc="Transforming SubstringsExtractor - substrings",
            ):
                X.loc[:, f"{col_pref}count-{substr}"] = X[column].str.count(substr) / (
                    X[column].str.len() if self.normalize else 1
                )

        if self.return_original_columns:
            return X
        return X.drop(columns=self.columns_to_apply)

In [16]:
from itertools import product

perms = [
    "".join(p)
    for i in range(1, 7)
    for p in product("ACTG", repeat=i)
    if len(p) < 5 or len(set(p)) > 2
]
ic(len(perms))
ic(perms[:10])

ic| len(perms): 4900
ic| perms[:10]: ['A', 'C', 'T', 'G', 'AA', 'AC', 'AT', 'AG', 'CA', 'CC']


['A', 'C', 'T', 'G', 'AA', 'AC', 'AT', 'AG', 'CA', 'CC']

In [17]:
%%capture --no-stdout --no-display
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

run_identification = f"{len(perms)} substrings"

# 1. step - define your pipeline
pipe = Pipeline(
    [
        (
            "fext_substr",
            SubstringsExtractor(
                substrings=perms,
            ),
        ),
        ("scale_u", UniversalScaler(scaler=MinMaxScaler())),
    ]
)

# 2. step - define your workflow
pipe_eval = PipelineEvaluator(
    pipeline=pipe,
    model=MLPClassifier(
        hidden_layer_sizes=(128, 32, 8),
        batch_size=32,
    ),
    evaluator=SequentialEvaluator(
        [
            ClassificationEvaluator(),
            UniversalEvaluator(metrics=[f1_score]),
        ]
    ),
)
# 3. step
pipe_eval.X_train = X_train.copy()
pipe_eval.y_train = y_train.copy()
pipe_eval.X_test = X_test.copy()
pipe_eval.y_test = y_test.copy()

result = pipe_eval.run()

results_for_comparison = results_for_comparison.append(
    pd.Series(result, name=run_identification),
)

09:18:56.245 Fitting pipeline
	Number of substrings BEFORE fit: 4900


Fitting SubstringsExtractor:   0%|          | 0/4900 [00:00<?, ?it/s]

	Number of substrings AFTER fit:  4900


Transforming SubstringsExtractor - columns:   0%|          | 0/1 [00:00<?, ?it/s]

Transforming SubstringsExtractor - substrings:   0%|          | 0/4900 [00:00<?, ?it/s]

09:30:48.776 Applying pipeline transformations


Transforming SubstringsExtractor - columns:   0%|          | 0/1 [00:00<?, ?it/s]

Transforming SubstringsExtractor - substrings:   0%|          | 0/4900 [00:00<?, ?it/s]

09:39:08.752 	Original shape:		(41290, 4900); 
		shape after removing NaNs: (41290, 4900).
09:39:09.637 	Original shape:		(30970, 4900); 
		shape after removing NaNs: (30970, 4900).
09:39:09.638 Fitting model
10:03:19.763 Predicting
10:03:21.244 Evaluating predictions
[[11587  3898]
 [ 3245 12240]] 
 76.93574426864707 % accuracy
 75.84582971867641 % precision of 1 classes
 79.04423635776558 % recall of 1 classes

              precision    recall  f1-score   support

           0       0.78      0.75      0.76     15485
           1       0.76      0.79      0.77     15485

    accuracy                           0.77     30970
   macro avg       0.77      0.77      0.77     30970
weighted avg       0.77      0.77      0.77     30970

f1_score:
	0.7741201024570723


In [18]:
results_for_comparison

accuracy_score  ...  f1_score
4900 substrings        0.769357  ...   0.77412

[1 rows x 5 columns]

| Dataset                          |   Accuracy |   F1 score |           |
|:---------------------------------|-----------:|-----------:|-----------:|
| human_enhancers_ensembl          |    81.8211 |    81.1213 | 340 substrings |
| human_enhancers_ensembl          |    83.9296 |    82.9747 | 1180 substrings | 	
| human_enhancers_ensembl          |    76.9357 |    77.412 | 4900 substrings - 1/3 train set | 	